torch/distributions/kl.py

```diff
@@ -462,6 +463,14 @@ def _kl_multivariatenormal_multivariatenormal(p, q):
     return half_term1 + 0.5 * (term2 + term3 - n)
 
 
+@register_kl(Multinomial, Multinomial)
+def _kl_multinomial_multinomial(p, q):
+    t = p.probs * (p.logits - q.logits)
+    t[(q.probs == 0).expand_as(t)] = inf
+    t[(p.probs == 0).expand_as(t)] = 0
+    return t.sum(-1)
```

Comment:
Maybe I'm missing something, but I don't actually completely understand why KL for multinomial-multinomial would be the same exact formula as categorical-categorical. 
Like, isn't the relationship between multinomial and categorical analogous to the relationship between binomial and Bernoulli?

Reply:
Analogous, but not identical -- the number of trials defines the binomial's shape, whereas the multinomial is defined simply by the probabilities. in this way categorical:multinomial are actually more convenience wrappers for each other than bernoulli:binomial.

Reply:
you can see, for example, in [this stackexchange question](https://math.stackexchange.com/questions/485810/kl-divergence-of-multinomial-distribution), that when you calculate the KL between two multinomials with the same number of categories, the number of trials doesn't effect the calculation

Reply:
From https://en.wikipedia.org/wiki/Multinomial_distribution:

> In [probability theory](https://en.wikipedia.org/wiki/Probability_theory), the multinomial distribution is a generalization of the [binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution). For example, it models the probability of counts for each side of a k-sided dice rolled n times. For n [independent](https://en.wikipedia.org/wiki/Statistical_independence) trials each of which leads to a success for exactly one of k categories, with each category having a given fixed success probability, the multinomial distribution gives the probability of any particular combination of numbers of successes for the various categories.

> When k is 2 and n is 1, the multinomial distribution is the [Bernoulli distribution](https://en.wikipedia.org/wiki/Bernoulli_distribution). When k is 2 and n is bigger than 1, it is the [binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution). When k is bigger than 2 and n is 1, it is the [categorical distribution](https://en.wikipedia.org/wiki/Categorical_distribution). The term "multinoulli" is sometimes used for the categorical distribution to emphasize this four-way relationship (so n determines the suffix, and k the prefix).

> The [Bernoulli distribution](https://en.wikipedia.org/wiki/Bernoulli_distribution) models the outcome of a single [Bernoulli trial](https://en.wikipedia.org/wiki/Bernoulli_trial). In other words, it models whether flipping a (possibly [biased](https://en.wikipedia.org/wiki/Fair_coin)) coin one time will result in either a success (obtaining a head) or failure (obtaining a tail). The [binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution) generalizes this to the number of heads from performing n independent flips (Bernoulli trials) of the same coin. The multinomial distribution models the outcome of n experiments, where the outcome of each trial has a [categorical distribution](https://en.wikipedia.org/wiki/Categorical_distribution), such as rolling a k-sided die n times.

Does this match your understanding?

Reply:
E.g. Multi(N, [0.5, 0.5]) = Bin(N, 0.5) and Multi(1, [0.5, 0.5]) = Bin(1, 0.5) = Categorical([0.5, 0.5]) = Bernoulli(0.5)